lms/security.py

```diff
@@ -87,7 +100,16 @@ def authenticated_userid(self, request):
         return identity.userid
 
     def identity(self, request):
-        lti_user = self._get_lti_user(request)
+        lti_user = None
+        try:
+            lti_user = self._get_lti_user(request)
+        except ValidationError as err:
+            # Keep a reference to the exception
+            # to later return a DeniedWithReason in `permits`
+            self._validation_error = err
+        else:
+            self._validation_error = None
```

Comment:
This branch is not necessary as this is instantiated per request but I reckon that not evident if reading this method alone.

Reply:
I wonder if the code would be simpler if `self._validation_error` could be avoided? Mutable state often makes if difficult to understand the flow of code. Could `permits()` below call `self._get_lti_user()` itself and catch the `ValidationError`,  instead of using `self._validation_error`?

```python
def permits(self, request, _context, permission):
    try:
        self._get_lti_user(request)
    except ValidationError as err:
        return DeniedWithValidationError(err)

    return _permits(self.identity(request), permission)
```

Then `identity()` would also do something similar:

```python
def identity(self, request):
    try:
        lti_user = self._get_lti_user(request)
    except ValidationError:
        return Identity("", [])

    permissions = [Permissions.LTI_LAUNCH_ASSIGNMENT, Permissions.API]

    if any(
        role in lti_user.roles.lower()
        for role in ["administrator", "instructor", "teachingassistant"]
    ):
        permissions.append(Permissions.LTI_CONFIGURE_ASSIGNMENT)

    return Identity(self._get_userid(lti_user), permissions, lti_user)
```

And maybe an `@cache` on the `_get_lti_user*()` functions could save any unnecessary work?